webapp/src/filelist.tsx

```diff
@@ -253,21 +254,102 @@ namespace custom {
             {show ? pxt.Util.concat(pkg.allEditorPkgs().map(p => this.filesWithHeader(p))) : undefined}
         </div>;
     }
 }
 
+interface GithubTreeItemProps extends ISettingsProps {
+    githubId: string;
+}
+
+interface GithubTreeItemState {
+    pushPulling?: boolean;
+}
+
+class GithubTreeItem extends sui.UIElement<GithubTreeItemProps, GithubTreeItemState> {
+    constructor(props: GithubTreeItemProps) {
+        super(props);
+        this.state = {};
+        this.handleClick = this.handleClick.bind(this);
+        this.handleButtonKeydown = this.handleButtonKeydown.bind(this);
+    }
+
+    private handleButtonKeydown(e: React.KeyboardEvent<HTMLElement>) {
+        e.stopPropagation();
+    }
+
+    private handleClick(e: React.MouseEvent<HTMLElement>) {
+        const { githubId } = this.props;
+        if (!githubId) {
+            this.createRepositoryAsync().done();
+        } else {
+            const gitf = pkg.mainEditorPkg().lookupFile("this/" + pxt.github.GIT_JSON);
+            this.props.parent.setSideFile(gitf);
+        }
+        e.stopPropagation();
+    }
+
+    private async createRepositoryAsync() {
+        pxt.tickEvent("github.filelist.create.start");
+        if (!pxt.github.token) await dialogs.showGithubLoginAsync();
+        if (!pxt.github.token) {
+            pxt.tickEvent("github.filelist.create.notoken");
+            return;
+        }
+
+        const repoid = await dialogs.showCreateGithubRepoDialogAsync(this.props.parent.state.projectName);
+        if (!repoid) return;
+
+        pxt.tickEvent("github.filelist.create.export");
+        core.showLoading("creategithub", lf("creating {0} repository...", pxt.github.parseRepoId(repoid).fullName))
+        try {
+            await workspace.exportToGithubAsync(this.props.parent.state.header, repoid);
+        } finally {
+            core.hideLoading("creategithub");
+        }
+        await this.props.parent.reloadHeaderAsync();
+    }
+
+    renderCore() {
+        const { githubId } = this.props;
+
+        // TODO: aria-selected, aria-label
+        //aria-selected={isActive}
+        //aria-label={isActive ? lf("{0}, it is the current opened file in the JavaScript editor", file.name) : file.name}
+
+
+        // todo: current branch
+        const ghid = pxt.github.parseRepoId(githubId);
+        const mainPkg = pkg.mainEditorPkg()
+        const meta: pkg.PackageMeta = ghid ? this.getData("open-pkg-meta:" + mainPkg.getPkgId()) : undefined;
+
+        return <a
+            key="github-status"
+            className="item"
+            onClick={this.handleClick}
+            tabIndex={0}
+            role="treeitem"
+            aria-selected="false"
+            aria-label="github status"
+            onKeyDown={sui.fireClickOnEnter}>
+            <i className="github icon" />
+            {ghid ? (ghid.project && ghid.tag ? `${ghid.project}#${ghid.tag}` : ghid.fullName) : lf("create GitHub repository")}
```

Comment:
nit: maybe ``(ghid.project && ghid.tag ? `${ghid.project}#${ghid.tag}` : ghid.fullName)`` could be pulled out into a const